### ISSUE TITLE
assumeReadType for reduction on distinct

### ISSUE THREAD

**Issue by jsantos17** [precog/quasar#3048]
Consider the query with distinct inside a reduction: `select state, sum(distinct(pop)) from zips group by state`.

<details>
<summary>QScript</summary>

```
QScript Mongo:
EquiJoin
├─ ShiftedRead(/jps/zips, ExcludeId)
├─ Reduce
│  ├─ ○
│  ├─ Bucket
│  │  ╰─ List
│  │     ╰─ MakeArray
│  │        ╰─ Guard
│  │           ├─ ○
│  │           ├─ Type(Obj(Map[], Some(Top)))
│  │           ├─ ProjectKey
│  │           │  ├─ ○
│  │           │  ╰─ Constant(Str(state))
│  │           ╰─ Undefined
│  ├─ Reducers
│  │  ╰─ Arbitrary
│  │     ╰─ Guard
│  │        ├─ ○
│  │        ├─ Type(Obj(Map[], Some(Top)))
│  │        ├─ ProjectKey
│  │        │  ├─ ○
│  │        │  ╰─ Constant(Str(state))
│  │        ╰─ Undefined
│  ╰─ Repair
│     ╰─ ReduceIndex(\/-(0))
├─ Reduce
│  ├─ Reduce
│  │  ├─ ○
│  │  ├─ Bucket
│  │  │  ╰─ List
│  │  │     ├─ Guard
│  │  │     │  ├─ Guard
│  │  │     │  │  ├─ ○
│  │  │     │  │  ├─ Type(Obj(Map[], Some(Top)))
│  │  │     │  │  ├─ ProjectKey
│  │  │     │  │  │  ├─ ○
│  │  │     │  │  │  ╰─ Constant(Str(pop))
│  │  │     │  │  ╰─ Undefined
│  │  │     │  ├─ Type(Int, Dec, or Interval)
│  │  │     │  ├─ Guard
│  │  │     │  │  ├─ ○
│  │  │     │  │  ├─ Type(Obj(Map[], Some(Top)))
│  │  │     │  │  ├─ ProjectKey
│  │  │     │  │  │  ├─ ○
│  │  │     │  │  │  ╰─ Constant(Str(pop))
│  │  │     │  │  ╰─ Undefined
│  │  │     │  ╰─ Undefined
│  │  │     ╰─ MakeArray
│  │  │        ╰─ Guard
│  │  │           ├─ ○
│  │  │           ├─ Type(Obj(Map[], Some(Top)))
│  │  │           ├─ ProjectKey
│  │  │           │  ├─ ○
│  │  │           │  ╰─ Constant(Str(state))
│  │  │           ╰─ Undefined
│  │  ├─ Reducers
│  │  │  ╰─ Arbitrary
│  │  │     ╰─ Guard
│  │  │        ├─ Guard
│  │  │        │  ├─ ○
│  │  │        │  ├─ Type(Obj(Map[], Some(Top)))
│  │  │        │  ├─ ProjectKey
│  │  │        │  │  ├─ ○
│  │  │        │  │  ╰─ Constant(Str(pop))
│  │  │        │  ╰─ Undefined
│  │  │        ├─ Type(Int, Dec, or Interval)
│  │  │        ├─ Guard
│  │  │        │  ├─ ○
│  │  │        │  ├─ Type(Obj(Map[], Some(Top)))
│  │  │        │  ├─ ProjectKey
│  │  │        │  │  ├─ ○
│  │  │        │  │  ╰─ Constant(Str(pop))
│  │  │        │  ╰─ Undefined
│  │  │        ╰─ Undefined
│  │  ╰─ Repair
│  │     ╰─ ConcatArrays
│  │        ├─ MakeArray
│  │        │  ╰─ ConcatArrays
│  │        │     ├─ MakeArray
│  │        │     │  ╰─ ReduceIndex(-\/(0))
│  │        │     ╰─ MakeArray
│  │        │        ╰─ ReduceIndex(-\/(1))
│  │        ╰─ MakeArray
│  │           ╰─ ReduceIndex(\/-(0))
│  ├─ Bucket
│  │  ╰─ List
│  │     ╰─ ProjectIndex
│  │        ├─ ProjectIndex
│  │        │  ├─ ○
│  │        │  ╰─ Constant(Int(0))
│  │        ╰─ Constant(Int(1))
│  ├─ Reducers
│  │  ╰─ Sum
│  │     ╰─ ProjectIndex
│  │        ├─ ○
│  │        ╰─ Constant(Int(1))
│  ╰─ Repair
│     ╰─ ReduceIndex(\/-(0))
├─ List
├─ Inner
╰─ ConcatMaps
   ├─ MakeMap
   │  ├─ Constant(Str(state))
   │  ╰─ LeftSide
   ╰─ MakeMap
      ├─ Constant(Str(1))
      ╰─ RightSide
```
</details>

The QScript still has unnecessary typechecks on `SrcHole`. `assumeReadType` is not traversing the reducers.

**Comment by jsantos17** [comment]
This is also happening for `EquiJoin`. Any idea on how hard this would be to implement @rintcius? It's breaking some plans (like in #3153)

**Comment by jsantos17** [comment]
This is an example of an `Equijoin` with redundant typechecks:

```
QScript Mongo (Prefer Projection):
EquiJoin
├─ Unreferenced
├─ ShiftedRead(/jps/zips, ExcludeId)
├─ ShiftedRead(/jps/zips2, ExcludeId)
├─ List
│  ╰─ tuple
│     ├─ Guard
│     │  ├─ ○
│     │  ├─ Type(Obj(Map[], Some(Top)))
│     │  ├─ ProjectKey
│     │  │  ├─ ○
│     │  │  ╰─ Constant(Str(pop))
│     │  ╰─ Undefined
│     ╰─ Guard
│        ├─ ○
│        ├─ Type(Obj(Map[], Some(Top)))
│        ├─ ProjectKey
│        │  ├─ ○
│        │  ╰─ Constant(Str(pop))
│        ╰─ Undefined
├─ Inner
╰─ Guard
   ├─ RightSide
   ├─ Type(Obj(Map[], Some(Top)))
   ├─ ProjectKey
   │  ├─ RightSide
   │  ╰─ Constant(Str(city))
   ╰─ Undefined
```

I think we might also want to simpliy `RightSide` here.

**Comment by rintcius** [comment]
`EquiJoin` is implemented as part of #2878.

`select state, sum(distinct(pop)) from zips group by state` does not compile correctly at the moment:

```
QScriptPlanningFailed(InternalError(ReifyBuckets: Expected to find a mappable region in 'qsu17 based on one of NonEmpty['qsu4].,None))
```

Looks similar to #3182 but apparently this is another case, because that issue is already fixed.
Reassigning to @wemrysi as he fixed the related issue. Can you reassign back to me once qscript is fixed?

**Comment by jsantos17** [comment]
Here's another query with non-elided `Guard`s in reductions: 

`select committer.login, count(*) from slamengine_commits`
